File: cpp/addOneRowToTree.cpp



//Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    TreeNode* addOneRow(TreeNode* root, int val, int depth) {
        if (depth == 1) {
            TreeNode* r = new TreeNode(val, root, nullptr);
            return r;
        }
        modify(root, val, 1, depth);
        return root;
    }   

    void modify(TreeNode* node, int val, int curdepth, int desireddepth) {
        if (curdepth == desireddepth - 1) {
            node->right = new TreeNode(val, nullptr, node->right);
            node->left = new TreeNode(val, node->left, nullptr);
            return;
        }
        else {
            if (node->right) modify(node->right, val, curdepth + 1, desireddepth);
            if (node->left) modify(node->left, val, curdepth + 1, desireddepth);
        }
    }
};

File: py/addOneRowToTree.py



# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def addOneRow(self, root: TreeNode, val: int, depth: int) -> TreeNode:
        if depth == 1:
            r = TreeNode(val, root)
            return r
        self.modify(root, val, 1, depth)
        return root
    
    def modify(self, node:TreeNode, val:int, cd:int, d:int):
        if cd == d - 1:
            node.left = TreeNode(val, node.left)
            node.right = TreeNode(val, None, node.right)
            return
        else:
            if node.left: self.modify(node.left, val, cd + 1, d)
            if node.right: self.modify(node.right, val, cd + 1, d)

